libslide/src/bignum/add.rs

```diff
@@ -0,0 +1,134 @@
+#![allow(clippy::suspicious_arithmetic_impl)]
+use crate::bignum::Bignum;
+use std::mem;
+use std::ops;
+
+impl ops::Add for Bignum {
+    type Output = Bignum;
+
+    fn add(self, rhs: Bignum) -> Bignum {
+        let is_neg: bool = rhs.is_neg && self.is_neg;
+
+        // this assumes the signs are both positive for now
+        let mut carry: u8 = 0;
+        let mut lhs_size: usize = self.dec.len();
+        let mut rhs_size: usize = rhs.dec.len();
+        let mut lhs_vec: Vec<u8>;
+        let mut rhs_vec: Vec<u8>;
+        let mut res: u8;
+
+        // 1. Handle Decimal
+
+        let mut res_dec: Vec<u8>;
+
+        // make lhs smaller vector
+        if lhs_size > rhs_size {
+            mem::swap(&mut lhs_size, &mut rhs_size);
+            lhs_vec = rhs.dec;
+            rhs_vec = self.dec;
+        } else {
+            lhs_vec = self.dec;
+            rhs_vec = rhs.dec;
+        }
+        res_dec = rhs_vec.clone(); 
+        let mut i = rhs_size as i8 - 1;
+        // add decimals that are unmatched by smaller vec
+        while i >= lhs_size as i8 {
+            res_dec[i as usize] = rhs_vec[i as usize];
+            i -= 1;
+        }
+
+        for i in (0..lhs_size).rev() {
+            res = lhs_vec[i] + rhs_vec[i] + carry;
+            res_dec[i] = res % 10;
+            carry = (res > 9) as u8
+        };
+
+        // 2. Handle Integer
+
+        lhs_size = self.int.len();
+        rhs_size = rhs.int.len();
+
+        let mut res_int: Vec<u8>;
+
+        // make lhs smaller vector
+        if lhs_size > rhs_size {
+            mem::swap(&mut lhs_size, &mut rhs_size);
+            lhs_vec = rhs.int;
+            rhs_vec = self.int;
+        } else {
+            lhs_vec = self.int;
+            rhs_vec = rhs.int;
+        }
+        res_int = rhs_vec.clone();
+
+        let diff: usize = rhs_size - lhs_size;
+
+        // compute the int vector (it actually does it backwards)
+        for i in (0..lhs_size).rev() {
+            res = lhs_vec[i] + rhs_vec[i+diff] + carry;
+            res_int[i+diff] = res % 10;
+            carry = (res > 9) as u8;
+        }
+
+        // fill with rest of the larger integer vector
+        // note we need to propogate carry here while in decimal we do not
+        for i in (0..diff).rev() {
+            res = rhs_vec[i] + carry;
+            println!("{}", res % 10);
+            res_int[i] = res % 10;
+            carry = (res > 9) as u8;
+        }
+
+        // add 1 if a carry is leftover
+        if carry == 1 {
+            res_int.insert(0, 1);
+        }
```

Comment:
Let's just push_back and flip how the integer is stored in the `BigNum`. insert is expensive because we have to shift all the elements.

Reply:
We can't flip how the integer is stored. We don't know where the decimal is located so to build from the decimal out we would have to flip it anyway.

Reply:
It would only slow performance because we would need to do reverses for both BigNums rather then just one shift here

Reply:
We only parse the number once, so flipping the decimal during construction is okay (just one O(n)) operation. But having to call insert everytime we do addition is way more expensive, because we now have an O(n) operation * # of times we call add. Plus, flipping is actually really only O(n/2) instructions.

Does this make sense? Basically what I mean is that for a number 123.456 we should store the int as 321. One reversal will be required during parsing and no reversals will be required during addition, at most one insertion to the end of the vector will be required.

Reply:
done.